// @shopify/discount-app-components/build/esm/components/ValueCard/ValueCard.js
import React from 'react';
import { Box, Card, BlockStack, Text, InlineStack, ButtonGroup, Button, TextField, InlineError, ChoiceList, Checkbox } from '@shopify/polaris';
import { useI18n } from '@shopify/react-i18n';
import { DiscountValueType, DiscountClass, PurchaseType } from '../../constants.js';
import { forcePositiveInteger } from '../../utilities/numbers.js';
import styles from './ValueCard.scss.js';
import CurrencyField from '../CurrencyField/CurrencyField.js';
import { AppliesTo } from '../AppliesTo/AppliesTo.js';

const FIXED_AMOUNT_VALUE_FIELD_ID = 'fixedAmountValueField';
const PERCENTAGE_VALUE_FIELD_ID = 'percentageValueField';
const ONCE_PER_ORDER_CHECKBOX_ID = 'oncePerOrderCheckbox';
const MAX_PERCENTAGE_LENGTH = 100;
const MIN_PERCENTAGE_LENGTH = 1;
function ValueCard({
  fixedAmountDiscountValue,
  percentageDiscountValue,
  discountValueType,
  purchaseType,
  oncePerOrder,
  discountClass,
  currencyCode,
  sellsSubscriptions,
  isCodeDiscount,
  eligibility,
  selectedItems,
  productSelector,
  collectionSelector
}) {
  const [i18n] = useI18n();
  const isPercentageDiscount = discountValueType.value === DiscountValueType.Percentage;
  const isProductDiscount = discountClass === DiscountClass.Product;
  const purchaseTypeChoices = [{
    value: PurchaseType.OneTimePurchase,
    label: i18n.translate('DiscountAppComponents.ValueCard.purchaseType.choices.oneTime')
  }, {
    value: PurchaseType.Subscription,
    label: i18n.translate('DiscountAppComponents.ValueCard.purchaseType.choices.subscription')
  }, {
    value: PurchaseType.Both,
    label: i18n.translate('DiscountAppComponents.ValueCard.purchaseType.choices.both')
  }];
  const handlePercentageValueChange = value => {
    return percentageDiscountValue.onChange(forcePositiveInteger(value));
  };
  const fixedAmountValueFloat = parseFloat(fixedAmountDiscountValue.value);
  const hasPercentageDiscountValueError = Boolean(percentageDiscountValue.error);
  return /*#__PURE__*/React.createElement(Box, {
    paddingBlockEnd: "400"
  }, /*#__PURE__*/React.createElement(Card, {
    padding: "400"
  }, /*#__PURE__*/React.createElement(BlockStack, {
    gap: "200"
  }, /*#__PURE__*/React.createElement(Text, {
    variant: "headingMd",
    as: "h2"
  }, i18n.translate('DiscountAppComponents.ValueCard.title'), ' '), /*#__PURE__*/React.createElement(InlineStack, {
    gap: "300",
    align: "start"
  }, /*#__PURE__*/React.createElement(ButtonGroup, {
    variant: "segmented"
  }, /*#__PURE__*/React.createElement(Button, {
    size: "large",
    pressed: isPercentageDiscount,
    onClick: () => discountValueType.onChange(DiscountValueType.Percentage)
  }, i18n.translate('DiscountAppComponents.ValueCard.percentageButton')), /*#__PURE__*/React.createElement(Button, {
    size: "large",
    pressed: !isPercentageDiscount,
    onClick: () => discountValueType.onChange(DiscountValueType.FixedAmount)
  }, i18n.translate('DiscountAppComponents.ValueCard.fixedAmountButton'))), /*#__PURE__*/React.createElement(Box, {
    width: "75%"
  }, !isPercentageDiscount && /*#__PURE__*/React.createElement(CurrencyField, {
    label: i18n.translate('DiscountAppComponents.ValueCard.discountValueLabel'),
    labelHidden: true,
    id: FIXED_AMOUNT_VALUE_FIELD_ID,
    value: fixedAmountDiscountValue.value,
    maxLength: 15,
    error: fixedAmountDiscountValue.error !== undefined,
    currencyCode: currencyCode,
    onChange: fixedAmountDiscountValue.onChange,
    onBlur: fixedAmountDiscountValue.onBlur,
    positiveOnly: true
  }), isPercentageDiscount && /*#__PURE__*/React.createElement(TextField, {
    autoComplete: "off",
    label: i18n.translate('DiscountAppComponents.ValueCard.discountValueLabel'),
    labelHidden: true,
    suffix: "%",
    value: percentageDiscountValue.value,
    onBlur: percentageDiscountValue.onBlur,
    maxLength: MAX_PERCENTAGE_LENGTH,
    minLength: MIN_PERCENTAGE_LENGTH,
    onChange: handlePercentageValueChange,
    error: hasPercentageDiscountValueError
  }), !isPercentageDiscount && fixedAmountDiscountValue.error && /*#__PURE__*/React.createElement("div", {
    className: styles.Error
  }, /*#__PURE__*/React.createElement(InlineError, {
    fieldID: FIXED_AMOUNT_VALUE_FIELD_ID,
    message: fixedAmountDiscountValue.error
  })), isPercentageDiscount && percentageDiscountValue.error && /*#__PURE__*/React.createElement("div", {
    className: styles.Error
  }, /*#__PURE__*/React.createElement(InlineError, {
    fieldID: PERCENTAGE_VALUE_FIELD_ID,
    message: percentageDiscountValue.error
  })))), sellsSubscriptions && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Text, {
    variant: "headingMd",
    as: "h2"
  }, i18n.translate('DiscountAppComponents.ValueCard.purchaseType.title')), isCodeDiscount ? /*#__PURE__*/React.createElement(ChoiceList, {
    title: i18n.translate('DiscountAppComponents.ValueCard.purchaseType.choiceListTitle'),
    titleHidden: true,
    choices: purchaseTypeChoices,
    selected: [purchaseType.value],
    onChange: values => purchaseType.onChange(values[0])
  }) : /*#__PURE__*/React.createElement("p", null, getPurchaseTypeWarning(discountClass, i18n))), isProductDiscount && !isPercentageDiscount && /*#__PURE__*/React.createElement(Checkbox, {
    id: ONCE_PER_ORDER_CHECKBOX_ID,
    label: i18n.translate('DiscountAppComponents.ValueCard.oncePerOrder'),
    checked: oncePerOrder.value,
    onChange: oncePerOrder.onChange,
    helpText: isNaN(fixedAmountValueFloat) ? i18n.translate('DiscountAppComponents.ValueCard.oncePerOrderHelpText') : i18n.translate('DiscountAppComponents.ValueCard.oncePerOrderHelpTextWithAmount', {
      fixedAmountValue: i18n.formatCurrency(Number(fixedAmountValueFloat), {
        currency: currencyCode,
        precision: 0,
        form: 'explicit'
      })
    })
  }), eligibility && selectedItems && (productSelector || collectionSelector) ? /*#__PURE__*/React.createElement(AppliesTo, {
    eligibility: eligibility,
    selectedItems: selectedItems,
    productSelector: productSelector,
    collectionSelector: collectionSelector
  }) : null)));
}
const getPurchaseTypeWarning = (discountClass, i18n) => {
  switch (discountClass) {
    case DiscountClass.Product:
      return i18n.translate('DiscountAppComponents.ValueCard.purchaseType.warning.product');
    case DiscountClass.Order:
      return i18n.translate('DiscountAppComponents.ValueCard.purchaseType.warning.order');
    case DiscountClass.Shipping:
      return i18n.translate('DiscountAppComponents.ValueCard.purchaseType.warning.shipping');
    default:
      return i18n.translate('DiscountAppComponents.ValueCard.purchaseType.warning.product');
  }
};

export { ValueCard };
