// @shopify/discount-app-components/build/esm/components/SummaryCard/components/Performance/Performance.js
import React from 'react';
import { useI18n } from '@shopify/react-i18n';
import { BlockStack, Text, List, Link } from '@shopify/polaris';
import { DiscountStatus, DiscountMethod } from '../../../../constants.js';

const CODE_DISCOUNT_ADMIN_REPORT_URL = `/reports/sales_by_discount`;
const I18N_SCOPE = {
  scope: 'DiscountAppComponents.SummaryCard.Performance'
};
function Performance({
  status,
  totalSales,
  hasReports,
  discountMethod,
  usageCount
}) {
  const [i18n] = useI18n();
  const isActiveOrExpired = status === DiscountStatus.Active || status === DiscountStatus.Expired;
  return /*#__PURE__*/React.createElement(BlockStack, {
    gap: "200"
  }, /*#__PURE__*/React.createElement(Text, {
    variant: "headingSm",
    as: "h3"
  }, i18n.translate('title', I18N_SCOPE)), status === DiscountStatus.Scheduled && /*#__PURE__*/React.createElement(Text, {
    as: "span",
    tone: "subdued"
  }, i18n.translate('notActive', I18N_SCOPE)), isActiveOrExpired && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(List, {
    type: "bullet"
  }, /*#__PURE__*/React.createElement(List.Item, null, i18n.translate('usageCount', I18N_SCOPE, {
    usageCount
  })), totalSales && /*#__PURE__*/React.createElement(List.Item, null, i18n.translate('totalSales', I18N_SCOPE, {
    totalSales: i18n.formatCurrency(Number(totalSales.amount), {
      currency: totalSales.currencyCode
    })
  }))), hasReports && discountMethod === DiscountMethod.Code && /*#__PURE__*/React.createElement("p", null, /*#__PURE__*/React.createElement(Link, {
    url: CODE_DISCOUNT_ADMIN_REPORT_URL
  }, i18n.translate('performanceLink', I18N_SCOPE)))));
}

export { Performance };
