// @shopify/discount-app-components/build/cjs/components/AppliesTo/AppliesTo.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var polaris = require('@shopify/polaris');
var reactI18n = require('@shopify/react-i18n');
var constants = require('../../constants.js');
var AppliesTo$1 = require('./AppliesTo.scss.js');
var SelectedItemsList = require('../SelectedItemsList/SelectedItemsList.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

function AppliesTo({
  eligibility,
  selectedItems,
  productSelector,
  collectionSelector
}) {
  const [i18n] = reactI18n.useI18n();
  React.useEffect(() => {
    if (!productSelector) {
      eligibility.onChange(constants.AppliesToEligibility.Collections);
    } else if (!collectionSelector) {
      eligibility.onChange(constants.AppliesToEligibility.Products);
    }
  }, [productSelector, collectionSelector, eligibility]);
  return /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, productSelector || collectionSelector ? /*#__PURE__*/React__default["default"].createElement(polaris.Box, {
    paddingBlockEnd: "400"
  }, /*#__PURE__*/React__default["default"].createElement(polaris.BlockStack, {
    gap: "400"
  }, /*#__PURE__*/React__default["default"].createElement(polaris.Text, {
    variant: "headingMd",
    as: "h2"
  }, i18n.translate('DiscountAppComponents.AppliesToCard.title')), /*#__PURE__*/React__default["default"].createElement(polaris.ChoiceList, {
    title: i18n.translate('DiscountAppComponents.AppliesToCard.choiceList.titleProducts'),
    disabled: !productSelector || !collectionSelector,
    titleHidden: true,
    choices: [{
      value: constants.AppliesToEligibility.Products,
      label: i18n.translate('DiscountAppComponents.AppliesToCard.choiceList.products')
    }, {
      value: constants.AppliesToEligibility.Collections,
      label: i18n.translate('DiscountAppComponents.AppliesToCard.choiceList.collections')
    }],
    selected: [eligibility.value],
    onChange: values => eligibility.onChange(values[0])
  }), eligibility.value === constants.AppliesToEligibility.Products && /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement("div", {
    className: AppliesTo$1["default"].SelectedItemsActivator
  }, productSelector), /*#__PURE__*/React__default["default"].createElement(SelectedItemsList.SelectedItemsList, {
    items: selectedItems.value.map(item => item),
    renderItem: item => /*#__PURE__*/React__default["default"].createElement("div", null, item.title),
    onRemoveItem: itemId => selectedItems.onChange(selectedItems.value.filter(item => item.id !== itemId))
  })), eligibility.value === constants.AppliesToEligibility.Collections && /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement("div", {
    className: AppliesTo$1["default"].SelectedItemsActivator
  }, collectionSelector), /*#__PURE__*/React__default["default"].createElement(SelectedItemsList.SelectedItemsList, {
    items: selectedItems.value.map(item => item),
    renderItem: item => /*#__PURE__*/React__default["default"].createElement("div", null, item.title),
    onRemoveItem: itemId => selectedItems.onChange(selectedItems.value.filter(item => item.id !== itemId))
  })))) : null);
}

exports.AppliesTo = AppliesTo;
