// @shopify/predicates/build/cjs/predicates.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function lengthMoreThan(length) {
  return input => input.length > length;
}
function lengthLessThan(length) {
  return input => input.length < length;
}
function isPositiveIntegerString(input) {
  return input !== '' && (input.match(/[^0-9]/g) || []).length === 0;
}
function isPositiveNumericString(input) {
  return input !== '' && (input.match(/[^0-9.,]/g) || []).length === 0;
}
function isNumericString(input) {
  return input !== '' && (input.match(/[^0-9.,-]/g) || []).length === 0;
}
function isEmpty(input) {
  return input === null || input === undefined || input.length === 0;
}
function isEmptyString(input) {
  return input === null || input === undefined || input.trim().length < 1;
}
function notEmpty(input) {
  return not(isEmpty)(input);
}
function notEmptyString(input) {
  return not(isEmptyString)(input);
}
function notNumericString(input) {
  return not(isNumericString)(input);
}
function not(fn) {
  return (...args) => !fn(...args);
}

exports.isEmpty = isEmpty;
exports.isEmptyString = isEmptyString;
exports.isNumericString = isNumericString;
exports.isPositiveIntegerString = isPositiveIntegerString;
exports.isPositiveNumericString = isPositiveNumericString;
exports.lengthLessThan = lengthLessThan;
exports.lengthMoreThan = lengthMoreThan;
exports.notEmpty = notEmpty;
exports.notEmptyString = notEmptyString;
exports.notNumericString = notNumericString;
