// @shopify/react-i18n/build/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var context = require('./context.js');
var manager = require('./manager.js');
var i18n = require('./i18n.js');
var hooks = require('./hooks.js');
var decorator = require('./decorator.js');
var types = require('./types.js');
var index = require('./constants/index.js');
var translate = require('./utilities/translate.js');
var interpolate = require('./utilities/interpolate.js');
var currencyCode = require('./currencyCode.js');
var currencyDecimalPlaces = require('./constants/currency-decimal-places.js');



exports.I18nContext = context.I18nContext;
exports.I18nManager = manager.I18nManager;
exports.I18n = i18n.I18n;
exports.useI18n = hooks.useI18n;
exports.withI18n = decorator.withI18n;
Object.defineProperty(exports, 'LanguageDirection', {
  enumerable: true,
  get: function () { return types.LanguageDirection; }
});
Object.defineProperty(exports, 'DateStyle', {
  enumerable: true,
  get: function () { return index.DateStyle; }
});
Object.defineProperty(exports, 'Weekday', {
  enumerable: true,
  get: function () { return index.Weekday; }
});
exports.memoizedNumberFormatter = translate.memoizedNumberFormatter;
exports.translate = translate.translate;
exports.DEFAULT_FORMAT = interpolate.DEFAULT_FORMAT;
exports.ERB_FORMAT = interpolate.ERB_FORMAT;
exports.MUSTACHE_FORMAT = interpolate.MUSTACHE_FORMAT;
Object.defineProperty(exports, 'CurrencyCode', {
  enumerable: true,
  get: function () { return currencyCode.CurrencyCode; }
});
exports.DEFAULT_DECIMAL_PLACES = currencyDecimalPlaces.DEFAULT_DECIMAL_PLACES;
exports.currencyDecimalPlaces = currencyDecimalPlaces.currencyDecimalPlaces;
